compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -30,6 +30,113 @@
 #include "il/Node_inlines.hpp"
 #include "infra/Bit.hpp"
 
+/**
+ * Helper that evaluates a child node.
+ * @param[in] childReg : the register of the child node being looked at
+ * @param[in] trgReg : the target register
+ * @param[in] child : the child node being looked at
+ * @param[in] parent : the parent of the child node being looked at
+ * @param[in] cg : codegenerator
+ */
+static inline void
+evaluateRegister(TR::Register **childReg, TR::Register *trgReg, TR::Node *child, TR::Node *parent, TR::CodeGenerator *cg)
+   {
+   int64_t value;
+	if (child->getOpCode().isLoadConst() && child->getRegister() == NULL)
+      {
+      value = child->getLongInt();
+      //if(constantIsUnsignedImm12(value))
+         //{
+         //TODO: Cover special cases for immediate loads
+         //}
+      //else
+         //{
+         *childReg = cg->allocateRegister();
+         loadConstant64(cg, parent, value, *childReg);
+         //}
+      }
+   else
+      {
+      *childReg = cg->evaluate(child);
+      }
+   if(child->getReferenceCount() == 1 && NULL == trgReg)
+      {
+      trgReg = *childReg;
+      }
+   }
+
+/**
+ * Generic helper that handles a number of similar binary operations.
+ * @param[in] node : calling node
+ * @param[in] regOp : the matching opcode of the caller
+ * @param[in] cg : codegenerator
+ * @return target register
+ */
+static inline TR::Register *
+genericBinaryEvaluator(TR::Node *node, TR::InstOpCode::Mnemonic regOp, TR::CodeGenerator *cg)
```

Comment:
If this is a binary evaluator, why isn't the number of children always "2" ?

Reply:
Our intent was to make this function able to handle other simple evaluators outside of the binary evaluators, though I'm not sure:
1) if that is something you wanted us to do, or if you would rather us stick to binary evaluators for now
2) how many unary or ternary operations actually exist that are simple and would fit this function (The ones that I have seen myself require extra steps).

Even if we remove this functionality, I think the function can easily be reincorporated later if similar patterns are showing up elsewhere in other evaluators. :)

Reply:
I think limiting this to just binary operators is sufficient for now.  Over-generalizing it to one or three children may overcomplicate it, and in fact cause us to miss opportunities or introduce bugs.